File: client/go/trailbase/transaction_api.go
package trailbase

import (
	"encoding/json"
	"fmt"
	"net/http"
)

type Operation struct {
	Type     string                 `json:"-"`
	ApiName  string                 `json:"api_name"`
	RecordID string                 `json:"record_id,omitempty"`
	Value    map[string]interface{} `json:"value,omitempty"`
}

func (o Operation) MarshalJSON() ([]byte, error) {
	var wrapper struct {
		Create *struct {
			ApiName string                 `json:"api_name"`
			Value   map[string]interface{} `json:"value"`
		} `json:"Create,omitempty"`
		Update *struct {
			ApiName  string                 `json:"api_name"`
			RecordID string                 `json:"record_id"`
			Value    map[string]interface{} `json:"value"`
		} `json:"Update,omitempty"`
		Delete *struct {
			ApiName  string `json:"api_name"`
			RecordID string `json:"record_id"`
		} `json:"Delete,omitempty"`
	}

	switch o.Type {
	case "Create":
		wrapper.Create = &struct {
			ApiName string                 `json:"api_name"`
			Value   map[string]interface{} `json:"value"`
		}{
			ApiName: o.ApiName,
			Value:   o.Value,
		}
	case "Update":
		wrapper.Update = &struct {
			ApiName  string                 `json:"api_name"`
			RecordID string                 `json:"record_id"`
			Value    map[string]interface{} `json:"value"`
		}{
			ApiName:  o.ApiName,
			RecordID: o.RecordID,
			Value:    o.Value,
		}
	case "Delete":
		wrapper.Delete = &struct {
			ApiName  string `json:"api_name"`
			RecordID string `json:"record_id"`
		}{
			ApiName:  o.ApiName,
			RecordID: o.RecordID,
		}
	}

	return json.Marshal(wrapper)
}

type TransactionRequest struct {
	Operations []Operation `json:"operations"`
}

type TransactionResponse struct {
	IDs []string `json:"ids"`
}

type TransactionBatch struct {
	client     Client
	operations []Operation
}

type ApiBatch struct {
	batch   *TransactionBatch
	apiName string
}

func (tb *TransactionBatch) API(apiName string) *ApiBatch {
	return &ApiBatch{
		batch:   tb,
		apiName: apiName,
	}
}

func (tb *TransactionBatch) Send() ([]string, error) {
	reqBody := TransactionRequest{
		Operations: tb.operations,
	}

	jsonData, err := json.Marshal(reqBody)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	resp, err := tb.client.do("POST", "/api/transactions/v1/execute", jsonData, []QueryParam{})
	if err != nil {
		return nil, fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
	}

	var response TransactionResponse
	decoder := json.NewDecoder(resp.Body)
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(&response); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	if response.IDs == nil {
		response.IDs = []string{} // Ensure non-nil slice
	}

	return response.IDs, nil
}

func (tb *TransactionBatch) addOperation(op Operation) {
	tb.operations = append(tb.operations, op)
}

func (ab *ApiBatch) Create(value map[string]interface{}) *TransactionBatch {
	ab.batch.addOperation(Operation{
		Type:    "Create",
		ApiName: ab.apiName,
		Value:   value,
	})
	return ab.batch
}

func (ab *ApiBatch) Update(recordID string, value map[string]interface{}) *TransactionBatch {
	ab.batch.addOperation(Operation{
		Type:     "Update",
		ApiName:  ab.apiName,
		RecordID: recordID,
		Value:    value,
	})
	return ab.batch
}

func (ab *ApiBatch) Delete(recordID string) *TransactionBatch {
	ab.batch.addOperation(Operation{
		Type:     "Delete",
		ApiName:  ab.apiName,
		RecordID: recordID,
	})
	return ab.batch
}

File: client/dotnet/trailbase/TransactionApi.cs
using System;
using System.Collections.Generic;
using System.Diagnostics.CodeAnalysis;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace TrailBase.Client {
  [JsonConverter(typeof(OperationJsonConverter))]
  public abstract class Operation {
    [JsonPropertyName("api_name")]
    public string ApiName { get; set; } = string.Empty;

    public static Operation Create(string apiName, Dictionary<string, object> value)
        => new CreateOperation { ApiName = apiName, Value = value };

    public static Operation Update(string apiName, string recordId, Dictionary<string, object> value)
        => new UpdateOperation { ApiName = apiName, RecordId = recordId, Value = value };

    public static Operation Delete(string apiName, string recordId)
        => new DeleteOperation { ApiName = apiName, RecordId = recordId };
  }

  public class CreateOperation : Operation {
    [JsonPropertyName("value")]
    public Dictionary<string, object> Value { get; set; } = new();
  }

  public class UpdateOperation : Operation {
    [JsonPropertyName("record_id")]
    public string RecordId { get; set; } = string.Empty;

    [JsonPropertyName("value")]
    public Dictionary<string, object> Value { get; set; } = new();
  }

  public class DeleteOperation : Operation {
    [JsonPropertyName("record_id")]
    public string RecordId { get; set; } = string.Empty;
  }

  public class OperationJsonConverter : JsonConverter<Operation> {
    [RequiresDynamicCode()]
    [RequiresUnreferencedCode()]
    public override Operation? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options) {
      if (reader.TokenType != JsonTokenType.StartObject)
        throw new JsonException();

      using (var doc = JsonDocument.ParseValue(ref reader)) {
        var root = doc.RootElement;

        if (root.TryGetProperty("Create", out var createElem)) {
          return JsonSerializer.Deserialize<CreateOperation>(createElem.GetRawText(), options);
        }
        else if (root.TryGetProperty("Update", out var updateElem)) {
          return JsonSerializer.Deserialize<UpdateOperation>(updateElem.GetRawText(), options);
        }
        else if (root.TryGetProperty("Delete", out var deleteElem)) {
          return JsonSerializer.Deserialize<DeleteOperation>(deleteElem.GetRawText(), options);
        }

        throw new JsonException("Unknown operation type");
      }
    }

    [RequiresUnreferencedCode()]
    [RequiresDynamicCode()]
    [RequiresDynamicCode()]
    public override void Write(Utf8JsonWriter writer, Operation value, JsonSerializerOptions options) {
      writer.WriteStartObject();

      switch (value) {
        case CreateOperation create:
          writer.WritePropertyName("Create");
          writer.WriteStartObject();
          writer.WriteString("api_name", create.ApiName);
          writer.WritePropertyName("value");
          JsonSerializer.Serialize(writer, create.Value, options);
          writer.WriteEndObject();
          break;

        case UpdateOperation update:
          writer.WritePropertyName("Update");
          writer.WriteStartObject();
          writer.WriteString("api_name", update.ApiName);
          writer.WriteString("record_id", update.RecordId);
          writer.WritePropertyName("value");
          JsonSerializer.Serialize(writer, update.Value, options);
          writer.WriteEndObject();
          break;

        case DeleteOperation delete:
          writer.WritePropertyName("Delete");
          writer.WriteStartObject();
          writer.WriteString("api_name", delete.ApiName);
          writer.WriteString("record_id", delete.RecordId);
          writer.WriteEndObject();
          break;

        default:
          throw new JsonException($"Unknown operation type: {value.GetType()}");
      }

      writer.WriteEndObject();
    }
  }

  public class TransactionRequest {
    [System.Text.Json.Serialization.JsonPropertyName("operations")]
    public List<Operation> Operations { get; set; } = new();
  }

  public class TransactionResponse {
    [System.Text.Json.Serialization.JsonPropertyName("ids")]
    public List<string> Ids { get; set; } = new();
  }

  public interface ITransactionBatch {
    IApiBatch Api(string apiName);
    Task<List<string>> SendAsync();
  }

  public interface IApiBatch {
    ITransactionBatch Create(Dictionary<string, object> record);
    ITransactionBatch Update(RecordId recordId, Dictionary<string, object> record);
    ITransactionBatch Delete(RecordId recordId);
  }

  public class TransactionBatch : ITransactionBatch {
    private readonly Client _client;
    private readonly List<Operation> _operations = new();

    public TransactionBatch(Client client) {
      _client = client;
    }

    public IApiBatch Api(string apiName) {
      return new ApiBatch(this, apiName);
    }

    [RequiresUnreferencedCode()]
    [RequiresDynamicCode()]
    [RequiresDynamicCode()]
    public async Task<List<string>> SendAsync() {
      var request = new TransactionRequest { Operations = _operations };
      var response = await _client.Fetch(
          "api/transaction/v1/execute",
          HttpMethod.Post,
          JsonContent.Create(request),
          null
      );

      string json = await response.Content.ReadAsStringAsync();
      var result = JsonSerializer.Deserialize<TransactionResponse>(json);

      return result?.Ids ?? new List<string>();
    }

    internal void AddOperation(Operation operation) {
      _operations.Add(operation);
    }
  }

  public class ApiBatch : IApiBatch {
    private readonly TransactionBatch _batch;
    private readonly string _apiName;

    public ApiBatch(TransactionBatch batch, string apiName) {
      _batch = batch;
      _apiName = apiName;
    }

    public ITransactionBatch Create(Dictionary<string, object> value) {
      _batch.AddOperation(Operation.Create(_apiName, value));
      return _batch;
    }

    public ITransactionBatch Update(RecordId recordId, Dictionary<string, object> value) {
      _batch.AddOperation(Operation.Update(_apiName, recordId.ToString(), value));
      return _batch;
    }

    public ITransactionBatch Delete(RecordId recordId) {
      _batch.AddOperation(Operation.Delete(_apiName, recordId.ToString()));
      return _batch;
    }
  }
}
